Refactor interactor

agirault opened the pull request:
- [x] Clean up unused functions in InteractorObserver
- [x] Improve switch to animation states in interactorStyle
- [x] Remove events not handled in Interactor
- [x] Consolidate setInteractor and setEnabled in InteractorObserver
- [x] Improve event state recording in the Interactor
- [x] Correctly store event positions in the interactor based on canvas
- [x] Fix issues in LineWidget and its representations
- [x] Pass event states as call data to interactor observers
- [x] Fix touch interaction

BREAKING CHANGES:
https://github.com/agirault/vtk-js/blob/0763417b649f9a5f5704d34a818f4831f5c9b39a/BREAKING_CHANGES.md#from-5x-to-6

agirault commented:
@jourdain @floryst @ken FYI, making some good progress on this.

The next step, now that the observers (styles + widget) and the interactor are better setup and organized, would be to organize data from events in state structures, and pass them down to the observers as calldata. Could do this in a separate PR, but that might create other breaking changes (maybe?) so if that was an issue I could keep working on that branch/PR.

jourdain commented on Sources/Interaction/Manipulators/MouseManipulator/index.js:
```diff
@@ -28,7 +29,7 @@ const DEFAULT_VALUES = {
   alt: false,
-  pinch: false,
+  scroll: false,
 };
```
As I told you, since you are refactoring it, pinch or scroll are not states to test against. They are more a mode that you want to be enabled against. Therefore, you can not keep the same naming. Maybe `enableMouseDrag`, `enableScroll` and `enablePinch`? Where by default all are false except `enableMouseDrag`...

agirault replied:
this PR is already huge, can we keep that for later? So that the behavior just doesn't change?

jourdain replied:
ok, do you want to merge to a branch `next` and create new PR. Then when we feel that `next` is ready we can merge it to master. Or do you rather merge that PR into master before doing any other work/PR?

jourdain commented:
It is hard to review and I trust your judgment on that. I've made a little comment to try to clarify some conflicting concept. 

Also make sure in your commit messages are properly tagged with each breaking changes so we will see them reported on the changelog. For that you can use `npm run commit`

agirault commented on Sources/Interaction/Widgets/AbstractWidget/index.js:
```diff
@@ -11,105 +8,41 @@ const { vtkErrorMacro } = macro;
-      }
-
-      if (model.currentRenderer) {
-        model.currentRenderer.removeViewProp(model.widgetRep);
```
Should we still hide it when disabling?

jourdain replied:
don't know what does VTK do?

floryst replied:
AFAIK VTK removes the prop like above.

Interestingly, LineWidget uses setEnabled(false) to turn off the widget, and then re-adds the representation into the scene. This has the effect of disabling widget interaction while keeping the widget in-scene. Maybe we can add a method like `setInteractionEnabled` so widgets can disable interactions, but keep the representation in the renderer?

agirault replied:
For the next round of things :)

agirault commented:
@martinken @floryst @jourdain Ready to go!

About 600 lines of code removed altogether, with:
- refactored/simplified observers
- event data passed through callbacks
- fixed position in interactor relative to the canvas
- fixed line widget
- fixed touch interactions
- all examples working
- passing tests

What I did not test was VR.

I listed the breaking changes in each commit (sometimes multiple per commits, hard to separate), and in the description of this PR above. @jourdain I also need to add them to a README, correct?

jourdain commented:
Just update https://github.com/Kitware/vtk-js/blob/master/BREAKING_CHANGES.md

I'll test the VR tomorrow.
